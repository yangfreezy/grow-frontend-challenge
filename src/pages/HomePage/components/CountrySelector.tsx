import React from 'react';
import { MenuItem, FormControl, InputLabel } from '@mui/material';
import Select, { SelectChangeEvent } from '@mui/material/Select';
import countryData from '../../../../assets/countries.json';

/*
    The countries.json file used here contains only the countries with functioning API responses.
    This info was not provided by the API so I had to scrape the API with a list of official ISO
    country codes to filter out the ones that 404'ed or errored out.
    
    I do realize this decouples our data with the API, but I decided I was okay with the tradeoff
    for a better user experience, since this would really only affect the least viewed countries respectively. 

    The alternatives I saw would have been to show a lot of countries to the user, but hundreds of them wouldn't work.
    Either that or somehow programatically determine which countries work with the API at runtime.
    Both of the above seemed to be an unfavorable tradeoff in either user experience and implementation complexity.
*/

interface Country {
    code: string;
    name: string;
}

type CountrySelectorProps = {
    country: Country;
    onChange: (val : Country) => void;
}

const stringMap: { [key: string]: string } = {};

const countryDictionary = countryData.reduce((acc, val) => {
    return {
        ...acc,
        [val.code]: val.name,
    }
}, stringMap);

const CountrySelector = ({ country, onChange } : CountrySelectorProps ) => {
    return (
        <FormControl data-testid="CountrySelector" sx={{ margin: 2.5 }}>
            <InputLabel id="country-quantity-helper-label">Country</InputLabel>
            <Select
                aria-label="Country Selector"
                labelId="country-quantity-select-label"
                id="country-quantity-select"
                value={country.code}
                data-testid="CountrySelectorSelect"
                label="Country"
                onChange={(e : SelectChangeEvent) : void => {
                    const { target: { value } } = e;
                    const country = { code: value, name: countryDictionary[value] };
                    onChange(country)
                }}
            >
            {
                countryData.map((country: Country) => (
                    <MenuItem data-testid={`MenuItem-${country.code}`} key={country.code} value={country.code}>{country.name}</MenuItem>
                ))
            }
            </Select>
          </FormControl>
    );
};

export default CountrySelector;
