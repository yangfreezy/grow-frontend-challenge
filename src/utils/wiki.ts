
import { Moment } from 'moment';
import { WikiPageViewMetrics } from "../types/wiki";

export const makeArticleTitleReadable = (article : string ) : string => article.split('_').join(' ');

// API does not come with it's own unique ID, so this is a workaround
export const getArticleId = (a: WikiPageViewMetrics) => a.article + '-' + a.project;

export const getWikiReqUrl = (date : Moment, countryCode: string) : string => {
    const dateToDisplay = date.format("YYYY/MM/DD");
    // Internationalization logic here for international products
    // I would have limited this with ?limit={numResults}, but the API doesn't seem to support limits
    return `https://wikimedia.org/api/rest_v1/metrics/pageviews/top-per-country/${countryCode}/all-access/${dateToDisplay}`;
};