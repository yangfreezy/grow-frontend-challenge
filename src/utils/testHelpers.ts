import { Country } from "../types/countries";
import countryData from '../../assets/countries.json';
import { WikiPageViewMetrics } from "../types/wiki";

export const getTestCountries = (num?: number) : Country[] => {
    if (!num) return countryData;
    return countryData.slice(0, num);
};

export const getTestArticles = (num: number) : WikiPageViewMetrics[] => {
    const arr = new Array(num);
    return arr.fill(0).map((a: WikiPageViewMetrics, i: number) => ({
        article: `testArticle${i}`,
        rank: i,
        views_ceil: i * 1000,
        project: 'en.wikipedia',
    })) 
};