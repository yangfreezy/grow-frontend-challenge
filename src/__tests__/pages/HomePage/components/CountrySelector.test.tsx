import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import CountrySelector from '../../../../pages/HomePage/components/CountrySelector';
import { getTestCountries } from '../../../../utils/testHelpers';

describe('CountrySelector', () => {
    let component: RenderResult;
    const mockOnChange = jest.fn();
    beforeEach(() => {
        component = render(<CountrySelector
            country={getTestCountries(1)[0]}
            onChange={mockOnChange}
        />);
    })

    it('renders', () => {
        expect(component).toBeTruthy();
    })

    it('matches the snapshot', () => {
        expect(component).toMatchSnapshot();
    })
})