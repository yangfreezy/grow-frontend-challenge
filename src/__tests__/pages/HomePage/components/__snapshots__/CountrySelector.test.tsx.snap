// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountrySelector matches the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiFormControl-root css-1km2w0t-MuiFormControl-root"
        data-testid="CountrySelector"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          id="country-quantity-helper-label"
        >
          Country
        </label>
        <div
          aria-label="Country Selector"
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
          data-testid="CountrySelectorSelect"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="country-quantity-select-label country-quantity-select"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="country-quantity-select"
            role="button"
            tabindex="0"
          >
            Albania
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            tabindex="-1"
            value="AL"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Country
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiFormControl-root css-1km2w0t-MuiFormControl-root"
      data-testid="CountrySelector"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        id="country-quantity-helper-label"
      >
        Country
      </label>
      <div
        aria-label="Country Selector"
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        data-testid="CountrySelectorSelect"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="country-quantity-select-label country-quantity-select"
          class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          id="country-quantity-select"
          role="button"
          tabindex="0"
        >
          Albania
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          tabindex="-1"
          value="AL"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Country
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
